### PR TITLE
Add screen lock feature in ExoPlayer

### LINKED ISSUE #77: Add "screen lock" feature
**Describe the feature you'd like**
Add a button in the video controls to lock the screen from accidental touch inputs. MX Player had this feature for years and Netflix recently added this too. Would love to have it in the Jellyfin app as well.

**Additional context**
Here are some screenshots from Netflix and MX Player:

Screenshot_20200830-163607

Screenshot_20200420_170917 0

### PR DESCRIPTION
**Pull request by CarlosOlivo**
Closes #77

### CODE REVIEW

**Comment by Maxr1998** [comment]
About the external player feature, I actually just pushed a PR that moves the external player selection to the native client settings :sweat_smile:
So, you can remove that code again ^^

**Comment by CarlosOlivo** [comment]
> About the external player feature, I actually just pushed a PR that moves the external player selection to the native client settings 😅
> So, you can remove that code again ^^

Nice timing.

**Comment by Maxr1998** [comment]
> Nice timing.

Yeah, sorry. We quickly spotted this after (actually even before) merging the PR yesterday, and decided to merge the player settings to a single location. I started working on it yesterday evening, but only finished it a short while ago.

File: app/src/main/java/org/jellyfin/mobile/utils/UIExtensions.kt [hunk 1]
```diff
@@ -46,6 +49,24 @@ fun Activity.disableFullscreen(keepStableLayout: Boolean = false) {
...
+@Suppress("DEPRECATION")
+fun Activity.lockOrientation() {
+    val display = windowManager.defaultDisplay
+    val rotation = display.rotation
+    val height: Int
+    val width: Int
+    val size = Point()
+    display.getSize(size)
+    height = size.y
+    width = size.x
```
Comment (Maxr1998) [inline app/src/main/java/org/jellyfin/mobile/utils/UIExtensions.kt:61]:
```suggestion
    val size = Point().also(display::getSize)
    val height = size.y
    val width = size.x
```

File: app/src/main/res/layout/activity_player.xml [hunk 2]
```diff
@@ -38,5 +38,20 @@
...
+
+        <androidx.appcompat.widget.AppCompatImageButton
+            android:id="@+id/unlock_screen_button"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_gravity="start|top"
+            android:layout_marginStart="@dimen/exo_player_controls_display_padding"
+            android:layout_marginTop="48dp"
+            android:layout_marginEnd="@dimen/exo_player_controls_display_padding"
+            android:background="@drawable/ic_screen_unlock_white_24dp"
```
Comment (Maxr1998) [inline app/src/main/res/layout/activity_player.xml:50]:
```suggestion
            android:src="@drawable/ic_screen_unlock_white_24dp"
```

Comment (Maxr1998) [inline app/src/main/res/layout/activity_player.xml:50]: You might want to add an actual background to ensure the button is always visible above the movie content, maybe a scrim/shadow, or a semi-transparent circle/rounded square.

File: app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt [hunk 3]
```diff
@@ -186,6 +206,24 @@ class PlayerActivity : AppCompatActivity() {
...
+                    unlockScreenButton.isVisible = true
+                    playerView.removeCallbacks(hideScreenButton)
+                    playerView.postDelayed(hideScreenButton, DEFAULT_CONTROLS_TIMEOUT_MS.toLong())
+                    return true
+                }
+            })
+        playerView.setOnTouchListener { _, event ->
+            gestureDetector.onTouchEvent(event)
+            true
+        }
```
Comment (Maxr1998) [inline app/src/main/java/org/jellyfin/mobile/player/PlayerActivity.kt:224]: Could you try to merge this with the `GestureDetector` above?

File: app/src/main/java/org/jellyfin/mobile/player/PlaybackMenus.kt [hunk 4]
```diff
@@ -20,6 +20,8 @@ class PlaybackMenus(private val activity: PlayerActivity) : PopupMenu.OnDismissL
     private val subtitlesButton: View = activity.findViewById(R.id.subtitles_button)
     private val audioStreamsButton: View = activity.findViewById(R.id.audio_streams_button)
     private val infoButton: View = activity.findViewById(R.id.info_button)
+    private val lockScreenButton: View = activity.findViewById(R.id.lock_screen_button)
+    private val unlockScreenButton: View = activity.findViewById(R.id.unlock_screen_button)
```
Comment (Maxr1998) [inline app/src/main/java/org/jellyfin/mobile/player/PlaybackMenus.kt:24]: You can move those above the subtitles button to follow the order of the layout (and while you're at it maybe also swap the audio and subtitles buttons, oops)

**Review by Maxr1998 — CHANGES_REQUESTED** [review]

**Comment by Maxr1998** [comment]
Thanks a lot for the PR, the code looks really good! Have some small issues I spotted, would appreciate if you could resolve them :innocent:

**Comment by CarlosOlivo** [comment]
> > Nice timing.
> 
> Yeah, sorry 🙈
> We quickly spotted this after (actually even before) merging the PR yesterday, and decided to merge the player settings to a single location. I started working on it yesterday evening, but only finished it a short while ago.

Well, I can say the same, I started working on this since yesterday and just finish it now, sorry for the spam.
Now the PR is as it should be from the beginning. I will check what I can correct.

**Comment by CarlosOlivo** [comment]
I don't know if it's an error or WIP, but ExoPlayer now don't report when the playback stops or ends, it keeps reporting "/Sessions /Playing/Progress", in the background, I don't know since when, in the v2.0.0-rc.3 version it works correctly, reporting "/Sessions/Playing/Stopped" at the end. @Maxr1998 @nielsvanvelzen

**Comment by nielsvanvelzen** [comment]
That might be a regression of #104. We'll need to look into that.

**Comment by Maxr1998** [comment]
> I don't know if it's an error or WIP, but ExoPlayer now don't report when the playback stops or ends, it keeps reporting "/Sessions /Playing/Progress", in the background, I don't know since when, in the v2.0.0-rc.3 version it works correctly, reporting "/Sessions/Playing/Stopped" at the end. @Maxr1998 @nielsvanvelzen 

Interesting, in my testing it successfully marked the title as played, but that may be because the progress was high enough (you can set in Jellyfin after what percentage played it's considered as watched, after all). I previously thought that the ending submission works by setting the progress to the runtime length, but there seems to be another API route then. I'll investigate.

**Review by Maxr1998 — APPROVED** [review]
I'll give this another test locally, and merge it afterwards :+1:

**Comment by Maxr1998** [comment]
Thanks again for the PR @CarlosOlivo, looking forward to see more contributions from you in the future!

**Pull request merged**